app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt

```diff
@@ -8,20 +8,144 @@
 
-import app.musikus.database.Nullable
-import app.musikus.database.daos.LibraryFolder
+import androidx.compose.runtime.Stable
+import androidx.compose.ui.graphics.Color
+import app.musikus.database.LibraryFolderWithItems
+import app.musikus.database.UUIDConverter
 import app.musikus.database.daos.LibraryItem
-import app.musikus.ui.library.DialogMode
-import app.musikus.ui.library.LibraryItemDialogUiEvent
-import app.musikus.ui.library.LibraryItemDialogUiState
-import app.musikus.ui.library.LibraryItemEditData
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
 import java.time.ZonedDateTime
 import java.util.UUID
-import kotlin.time.Duration
 
 
+// TODO UIText for all Strings?
 
+@Stable
+enum class SessionPausedResumedState {
+    RUNNING,
+    PAUSED
+}
+
+/** Specific one-time events triggering some UI action in ActiveSession */
+@Stable
+data class ActiveSessionEventStates (
+    val sessionSaved: Boolean = false,
+    val sessionDiscarded: Boolean = false,
+)
+
+@Stable
+/** Main UI State */
 data class ActiveSessionUiState(
+    val topBarUiState: StateFlow<ActiveSessionTopBarUiState> = MutableStateFlow(ActiveSessionTopBarUiState()),
+    val mainContentUiState: StateFlow<MainContentUiState> = MutableStateFlow(MainContentUiState()),
+    val newItemSelectorUiState: StateFlow<NewItemSelectorUiState> = MutableStateFlow(NewItemSelectorUiState()),
+    val toolsUiState: StateFlow<ActiveSessionToolsUiState> = MutableStateFlow(ActiveSessionToolsUiState()),
+    val dialogVisibilities: StateFlow<DialogVisibilities> = MutableStateFlow(DialogVisibilities()),
+)
+
+data class DialogVisibilities(
+    val newItemSelectorVisible: Boolean = false,
+    val finishDialogVisible: Boolean = false,
+    val discardDialogVisible: Boolean = false,
+    val createItemDialogVisible: Boolean = false,
+    val createFolderDialogVisible: Boolean = false,
+)
+
+@Stable
+data class MainContentUiState(
+    val timerUiState: StateFlow<ActiveSessionTimerUiState> = MutableStateFlow(ActiveSessionTimerUiState()),
+    val currentItemUiState: StateFlow<ActiveSessionCurrentItemUiState> = MutableStateFlow(ActiveSessionCurrentItemUiState()),
+    val pastSectionsUiState: StateFlow<ActiveSessionCompletedSectionsUiState> = MutableStateFlow(ActiveSessionCompletedSectionsUiState()),
+    val endDialogUiState: StateFlow<ActiveSessionEndDialogUiState> = MutableStateFlow(ActiveSessionEndDialogUiState())
+)
+
+@Stable
+data class ActiveSessionTopBarUiState(
+    val visible: Boolean = false,
+    val pauseButtonAppearance: SessionPausedResumedState = SessionPausedResumedState.RUNNING,
+)
+
+@Stable
+data class ActiveSessionTimerUiState(
+    val timerText: String = "00:00", // Prevents size change animation to trigger when opening the screen
+    val subHeadingAppearance: SessionPausedResumedState = SessionPausedResumedState.RUNNING,
+    val subHeadingText: String = ""
+)
+
+@Stable
+data class ActiveSessionCurrentItemUiState(
```

Comment:
If we simply make the `ActiveSessionCurrentItemUiState` nullable inside `MainContentUiState`, we can save ourselves the `visbile` property and the awkward default values

Reply:
Yes, with visibility it's always difficult because IMO visibility is part of the general UI state but actually not part of the element itself whether it is shown or not so yeah might be a solution to use the Nullable type... Although I don't agree with the default values, see above.

Reply:
I tried to make them nullable (via `?`) but then in Compose, smart-casting is not possible anymore, so is early returning when the state is `null`. That means we would always have to define unnecessary fallbacks for `null` even though we don't use the function.

Wrapping the States with `Nullable<>` makes them smart-castable in Composables but then the StateFlow `combine` functions complain about mismatched types because the compiler interprets them as `Nullable<out StateClass>>` rather than `Nullable<StateClass>` 🤔

![image](https://github.com/matthiasemde/musikus-android/assets/28269133/7438c55b-c970-48ba-a996-75c5142c4d02)

(this also happens with `initialValue = Nullable(null)`)